strawberry/permission.py

```diff
@@ -1,21 +1,125 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, Awaitable, Optional, Union
+import abc
+from inspect import iscoroutinefunction
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Awaitable,
+    Callable,
+    List,
+    Optional,
+    Type,
+    Union,
+    cast,
+)
+
+from strawberry.exceptions import StrawberryGraphQLError
+from strawberry.extensions import FieldExtension
+from strawberry.schema_directive import Location, StrawberrySchemaDirective
+from strawberry.utils.await_maybe import await_maybe
+from strawberry.utils.cached_property import cached_property
 
 if TYPE_CHECKING:
-    from strawberry.types.info import Info
+    from graphql import GraphQLError, GraphQLErrorExtensions
+
+    from strawberry.field import StrawberryField
+    from strawberry.types import Info
 
 
-class BasePermission:
+class BasePermission(abc.ABC):
     """
     Base class for creating permissions
     """
 
     message: Optional[str] = None
 
+    error_extensions: Optional[GraphQLErrorExtensions] = None
+
+    error_class: Type[GraphQLError] = StrawberryGraphQLError
+
+    _schema_directive: Optional[StrawberrySchemaDirective] = None
+
+    @abc.abstractmethod
     def has_permission(
         self, source: Any, info: Info, **kwargs
     ) -> Union[bool, Awaitable[bool]]:
         raise NotImplementedError(
             "Permission classes should override has_permission method"
         )
+
+    def get_error(self) -> GraphQLError:
```

Comment:
Wouldn't just a `handle_error` (or `handle_no_permission` be enough? And this function raise the error itself?

For my use case I can just use FieldExtension and override `resolve` directly, so it is not a major issue if you prefer to leave it like this :)

Reply:
This way we're leaving things open for typed errors using the same interface in the future. I'm fine with both ways of handling it 🙂

Reply:
After thinking it through again, it seems to be cleaner to separate typed error permissions from normal GQLError permissions. If that changes, we can revisit that later.